ci-bench/src/bench_lib/cachegrind.rs

```diff
@@ -0,0 +1,237 @@
+    /// actually running one of the benchmarks
+    ///
+    /// This count is subtracted from benchmark results, to reduce noise
+    overhead_instructions: u64,
+}
+
+impl CachegrindRunner {
+    /// Returns a new cachegrind-based benchmark runner
+    pub fn new(executable: String) -> anyhow::Result<Self> {
+        Self::ensure_cachegrind_available()?;
+        let arch = Self::get_arch()?;
+
+        // We don't care about the side here, so let's use `Server` just to choose something
+        let overhead_instructions = Self::run_bench_side(
+            &arch,
+            &executable,
+            u32::MAX,
+            Side::Server,
+            "calibration",
+            Stdio::piped(),
+            Stdio::piped(),
+        )?
+        .wait_and_get_instr_count()
+        .context("Unable to count overhead instructions")?;
+
+        Ok(CachegrindRunner {
+            arch,
+            executable,
+            overhead_instructions,
+        })
+    }
+
+    /// Runs the benchmark at the specified index and returns the instruction counts for each side
+    pub fn run_bench(
+        &self,
+        benchmark_index: u32,
+        bench: &Benchmark,
+    ) -> anyhow::Result<InstructionCounts> {
+        // The server and client are started as child processes, and communicate with each other
+        // through stdio.
+
+        let mut server = Self::run_bench_side(
+            &self.arch,
+            &self.executable,
+            benchmark_index,
+            Side::Server,
+            &bench.name_with_side(Side::Server),
+            Stdio::piped(),
+            Stdio::piped(),
+        )
+        .context("server side bench crashed")?;
+
+        let client = Self::run_bench_side(
+            &self.arch,
+            &self.executable,
+            benchmark_index,
+            Side::Client,
+            &bench.name_with_side(Side::Client),
+            Stdio::from(server.process.stdout.take().unwrap()),
+            Stdio::from(server.process.stdin.take().unwrap()),
+        )
+        .context("client side bench crashed")?;
+
+        let counts = InstructionCounts {
+            server: server.wait_and_get_instr_count()?,
+            client: client.wait_and_get_instr_count()?,
+        };
+
+        let overhead_counts = InstructionCounts {
+            server: self.overhead_instructions,
+            client: self.overhead_instructions,
+        };
+
+        Ok(counts - overhead_counts)
+    }
+
+    /// Returns the CPU architecture
+    fn get_arch() -> anyhow::Result<String> {
+        let output = Command::new("uname")
+            .arg("-m")
+            .stdout(Stdio::piped())
+            .output()
+            .context("Failed to run `uname` to determine CPU architecture.")?;
+
+        let arch = String::from_utf8(output.stdout)
+            .context("`-uname -m` returned invalid unicode.")?
+            .trim()
+            .to_owned();
+
+        Ok(arch)
+    }
```

Comment:
Could this be replaced with https://doc.rust-lang.org/std/env/consts/constant.ARCH.html ?

Reply:
I'm afraid this could break in unexpected ways, because `uname -m` does not always use the same string representation of an architecture as the `ARCH` constant (see [this list](https://stackoverflow.com/a/45125525/2110623) for details).

But there is an even better option! I just discovered that the `arch` argument to `setarch` is optional since the 2.33 release (in 2018), so we can remove the `get_arch` function in its entirety.

Reply:
> But there is an even better option! I just discovered that the arch argument to setarch is optional since the 2.33 release (in 2018), so we can remove the get_arch function in its entirety.

Ah perfect, that sounds even better. Nice catch.